### ISSUE TITLE
Is ok to do some code optimization in NNVM?

### ISSUE THREAD

**Issue by dianyancao** [dmlc/mxnet#3105]
I just hear NNVM today,and I think it maybe neural network virtual machine?

Some code optimizations example:
Dropout:
We can first generate the mask for dropout and do required computations on input according to mask=1 directly 
BatchNorm:
In test phase,when use use_global_stats=True in BatchNorm,we can associate the (moving_mean,moving_var,gamma,beta) with (weight,bias) in last operator by associative law.

We can do it just by auto c++ code generation and optimization when some patterns occurd,and not need to change the interfaces for users by hide the optimizations for users.

This maybe a good feature I think,regard for your reply.

**Comment by dianyancao** [comment]
After generate the optimized code,we can compile it to a DLL and use Executer to excute the operations defined in the DLL.

**Comment by piiswrong** [comment]
code generation is an important feature. We were considering halide.
But don't have the man power to do it right now. You are certainly welcome to work on this.
If you do want to pitch in, please keep us updated on your design decisions.

**Comment by dianyancao** [comment]
@piiswrong I have no experiments on compiler and code generation.But I would like to learn to use Halide and LLVM for doing this.
Do you plan to support high order functions in operators by code generation?
Such that we can use imperative operations such as ```if for``` to define a symbol variable.  
And by code generation,the optimization similarly to above can be done by composite some successive operators into one operator then we can do a lot of optimizations in it.

**Comment by piiswrong** [comment]
NNVM itself doesn't include rewrite a operators, but it makes it easier for you to add a parallel track of Halide operators that can be optionally turned on.

**Comment by tqchen** [comment]
https://github.com/dmlc/nnvm it is now opensourced here
